File: chess.ts/field.test.ts
import { expect, test } from '@jest/globals'
import { Field, readField } from "./field"

test('game', () => {
    expect(new Field(1,1).show()).toBe('a1')
    expect(new Field(3,4).show()).toBe('c4')
    expect(readField('c4').col).toBe(3)
    expect(readField('c4').row).toBe(4)
    expect(readField('C4').col).toBe(3)
    expect(readField('C4').row).toBe(4)
    expect(new Field(3,4).relative([1,1]).col).toBe(4)
    expect(new Field(3,4).relative([1,1]).row).toBe(5)
    expect(new Field(3,4).relativeFields([[1,1],[3,4]])[0].col).toBe(4)
    expect(new Field(3,4).relativeFields([[1,1],[3,4]])[0].row).toBe(5)
    expect(new Field(3,4).relativeFields([[1,1],[3,4]])[1].col).toBe(6)
    expect(new Field(3,4).relativeFields([[1,1],[3,4]])[1].row).toBe(8)
    expect(readField('c4').isLastRow('Black')).toBe(false)
    expect(readField('d8').isLastRow('Black')).toBe(false)
    expect(readField('d8').isLastRow('White')).toBe(true)
    expect(readField('d8').isValid()).toBe(true)
    expect(readField('a0').isValid()).toBe(false)
    expect(readField('i8').isValid()).toBe(false)
    expect(readField('h9').isValid()).toBe(false)
})

File: chess.py/field_test.py
import unittest

from color import Color
from field import Field


class TestField(unittest.TestCase):
    def test_relative(self):
        self.assertEqual(Field(6, 7).relative((-2, -4)), Field(col=4, row=3))
        self.assertEqual(Field(3, 5).relative((9, 10)), Field(col=12, row=15))

    def test_relative_fields(self):
        self.assertEqual(
            Field(6, 7).relative_fields([(1, 0), (1, 1), (-2, -3), (9, 9)]),
            [
                Field(col=7, row=7),
                Field(col=7, row=8),
                Field(col=4, row=4),
                Field(col=15, row=16),
            ],
        )

    def test_is_last_row(self):
        self.assertEqual(Field(8, 8).is_last_row(Color.White), True)
        self.assertEqual(Field(7, 8).is_last_row(Color.White), True)
        self.assertEqual(Field(7, 8).is_last_row(Color.Black), False)
        self.assertEqual(Field(7, 1).is_last_row(Color.Black), True)
        self.assertEqual(Field(2, 2).is_last_row(Color.Black), False)

    def test_valid(self):
        self.assertEqual(Field(2, 2).is_valid(), True)
        self.assertEqual(Field(0, 2).is_valid(), False)
        self.assertEqual(Field(2, 0).is_valid(), False)
        self.assertEqual(Field(2, 9).is_valid(), False)
        self.assertEqual(Field(9, 2).is_valid(), False)
